File: leetcode/0394-decode-string/Solution.java
import java.util.Deque;
import java.util.ArrayDeque;

class Solution {
    public String decodeString(String s) {
        Deque<String> stringStack = new ArrayDeque<>();
        Deque<Integer> numStack = new ArrayDeque<>();
        StringBuilder current = new StringBuilder();
        int currentNum = 0;

        for (char ch : s.toCharArray()) {
            if (Character.isDigit(ch)) {
                currentNum = currentNum * 10 + (ch - '0');
            } else if (ch == '[') {
                stringStack.push(current.toString());
                numStack.push(currentNum);
                current = new StringBuilder();
                currentNum = 0;
            } else if (ch == ']') {
                int num = numStack.pop();
                String prev = stringStack.pop();
                StringBuilder repeated = new StringBuilder(prev);
                for (int i = 0; i < num; i++) {
                    repeated.append(current);
                }
                current = repeated;
            } else {
                current.append(ch);
            }
        }

        return current.toString();
    }
}

File: leetcode/0394-decode-string/solution.cpp
#include <string>
#include <stack>
#include <cctype>
using namespace std;

class Solution {
public:
    string decodeString(string s) {
        stack<pair<string, int>> st;  // (previous_string, repeat_count)
        string current;
        int currentNum = 0;

        for (char ch : s) {
            if (isdigit(ch)) {
                currentNum = currentNum * 10 + (ch - '0');
            } else if (ch == '[') {
                st.push({current, currentNum});
                current = "";
                currentNum = 0;
            } else if (ch == ']') {
                auto [prev, num] = st.top();
                st.pop();
                string repeated = prev;
                for (int i = 0; i < num; i++) {
                    repeated += current;
                }
                current = repeated;
            } else {
                current += ch;
            }
        }

        return current;
    }
};
